src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -666,5 +580,223 @@ void BaseQNAcceleration::writeInfo(
   }
   _infostringstream << std::flush;
 }
+
+void BaseQNAcceleration::concatenateCouplingData(const DataMap &cplData)
+{
+
+  // Needs to be called every iteration, since the time window size can vary with participant first
+  moveTimeGridToNewWindow(cplData);
+
+  /// If not reduced quasi-Newton then sample the residual of data in dataIDs to the corresponding time grid in _timeGrids and concatenate everything into a long vector
+  Eigen::Index offset = 0;
+  for (int id : _primaryDataIDs) {
+    auto            waveform        = cplData.at(id)->timeStepsStorage();
+    Eigen::Index    dataSize        = cplData.at(id)->getSize();
+    Eigen::VectorXd primaryTimeGrid = _primaryTimeGrids.at(id);
+
+    for (int i = 0; i < primaryTimeGrid.size(); i++) {
+
+      Eigen::VectorXd primaryData    = waveform.sample(primaryTimeGrid(i));
+      Eigen::VectorXd oldPrimaryData = cplData.at(id)->getPreviousValuesAtTime(primaryTimeGrid(i));
+
+      PRECICE_ASSERT(_primaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+      PRECICE_ASSERT(_oldPrimaryValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");
+
+      for (Eigen::Index i = 0; i < dataSize; i++) {
+        _primaryValues(i + offset)    = primaryData(i);
+        _oldPrimaryValues(i + offset) = oldPrimaryData(i);
+      }
+      offset += dataSize;
+    }
+  }
+
+  /// Sample all the data to the corresponding time grid in _timeGrids and concatenate everything into a long vector
+  offset = 0;
+
+  for (int id : _dataIDs) {
+    auto            waveform = cplData.at(id)->timeStepsStorage();
+    Eigen::Index    dataSize = cplData.at(id)->getSize();
+    Eigen::VectorXd timeGrid = _timeGrids.at(id);
+
+    for (int i = 0; i < timeGrid.size(); i++) {
+
+      Eigen::VectorXd data    = waveform.sample(timeGrid(i));
+      Eigen::VectorXd oldData = cplData.at(id)->getPreviousValuesAtTime(timeGrid(i));
+
+      PRECICE_ASSERT(_values.size() >= offset + dataSize, "the values were not initialized correctly");
+
+      for (Eigen::Index i = 0; i < dataSize; i++) {
+        _values(i + offset)    = data(i);
+        _oldValues(i + offset) = oldData(i);
+      }
+      offset += dataSize;
+    }
+  }
+}
```

Comment:
The blocks `for (int id : _primaryDataIDs)` and `for (int id : _dataIDs)` are almost identical. I think you could remove some duplication by introducing a function:
```suggestion
void BaseQNAcceleration::concatenateCouplingData(const DataMap &cplData)
{
  // Needs to be called every iteration, since the time window size can vary with participant first
  moveTimeGridToNewWindow(cplData);

    /// If not reduced quasi-Newton then sample the residual of data in dataIDs to the corresponding time grid in _timeGrids and concatenate everything into a long vector  
  doConcatenate(_primaryValues, _oldPrimaryValues, _primaryDataIDs, _primaryTimeGrids);
  
  /// Sample all the data to the corresponding time grid in _timeGrids and concatenate everything into a long vector
  doConcatenate(_values, _oldValues, _dataIDs, _timeGrids);
}
```
I think this also makes sense since you forgot the assertion `PRECICE_ASSERT(_oldValues.size() >= offset + dataSize, "the primaryValues were not initialized correctly");` in the second block. This cannot happen with a function that easily.